### PR TITLE
fio: add new job file for testing on local storage and enhancements for fio tests

### PR DESCRIPTION
**Pull request by lukazlim**
- Add 'format: n' parameter for fs will skip formatting of target partition
- Add support for fio tests to run on null block device
- Set fio testfile directory to /lkp/benchmarks/fio if target partition is local storage (root partition '/')

To support fio tests on local storage and null block device with new fio job file fio-basic-local-disk.yaml

### CODE REVIEW

File: setup/fio-setup-basic [hunk 1]
```diff
@@ -179,11 +179,15 @@ parse_numa_mem_policy
...
 no=0
+
 for storage in $storages; do
 	if [ "$raw_disk" = 1 ]; then
 		storage_setup="filename=$storage"
 	else
-		storage_setup="directory=$storage"
+		storage_setup="\
+directory=$storage
+filename=fio_testfile
```
Comment (rli9) [inline setup/fio-setup-basic:203]: do you know what is the default filename if it is not specified before? and this is not related to format logic, thus help split this to a single commit.

BTW: you don't need create a new pr, you can just force push to the pr which requires change.

File: setup/fio-setup-basic [hunk 2]
```diff
@@ -184,10 +184,14 @@ for storage in $storages; do
 	if [ "$raw_disk" = 1 ]; then
 		storage_setup="filename=$storage"
 	else
-		storage_setup="\
+		if [[ "$storage" =~ "/null" ]]; then
```
Comment (rli9) [inline setup/fio-setup-basic:192]: the script is bin/sh, it may not support double brackets, and =~.

File: setup/fio-setup-basic [hunk 3]
```diff
@@ -175,6 +175,10 @@ __numa_cpu_nodes()
 	esac
 }
 
+is_mount_on_root() {
```
Comment (rli9) [inline setup/fio-setup-basic:178]: help move { to next line

File: setup/fio-setup-basic [hunk 4]
```diff
@@ -175,6 +175,10 @@ __numa_cpu_nodes()
...
+	[[ $(df | grep $1 | awk '{print $6}') == '/' ]]
```
Comment (rli9) [inline setup/fio-setup-basic:179]: i wonder whether this syntax can work in sh (which is not necessary bash). Also does awk need be installed through deps?

Comment (lukazlim) [reply setup/fio-setup-basic:179]: no additional installation needed for awk but will use grep instead to maximize compatibility

File: setup/fio-setup-basic (hunk 1 above)
Comment (rli9) [reply setup/fio-setup-basic:211]: what is the default behavior if you don't specify filename here? Which filename it will use?

File: setup/fio-setup-basic [hunk 5]
```diff
@@ -175,6 +175,12 @@ __numa_cpu_nodes()
 	esac
 }
 
+is_mount_on_root()
+{
```
Comment (rli9) [inline setup/fio-setup-basic:179]: can you update commit subject to be something like below to reflect the relation with root?

fio: run test on /lkp/benchmarks/fio when storage being mounted on root

File: setup/fio-setup-basic (hunk 1 above)
Comment (lukazlim) [reply setup/fio-setup-basic:211]: oh, it will create a file named 'task_0.0.0' with size of specified test size in the provided directory
then running different fio tests will create task_0.1.0 , task_0.2.0 , task_0.3.0 , and so on if directory provided is the same

Comment (rli9) [reply setup/fio-setup-basic:211]: I'm afraid this impact other users if they depend on default behavior to have multiple results. From the patch, multiple runs write to single file fio_testfile, am i right?

If this is specific to your needs, can it be some parameter in job yaml?

Comment (lukazlim) [reply setup/fio-setup-basic:211]: Yes, multiple runs will write to the single file fio_testfile to fulfill specific requirement
I will add a parameter to determine if fio_testfile needs to be specified or not

File: jobs/fio-basic-local-disk.yaml [hunk 6]
```diff
@@ -32,5 +32,6 @@ fio-setup-basic:
     - 1
     - 32
   test_size: 1g
+  testfile: y
```
Comment (rli9) [inline jobs/fio-basic-local-disk.yaml:35]: this patch need be merged with early patch, to make each commit of the series works as expected.

File: setup/fio-setup-basic [hunk 7]
```diff
@@ -185,13 +191,18 @@ for storage in $storages; do
 	if [ "$raw_disk" = 1 ]; then
 		storage_setup="filename=$storage"
 	else
-		if [ "$testfile" = 'y' ]; then
-			storage_setup="\
+		dev=/dev/$(basename $storage)
+		if is_null_blk $dev; then
+			storage_setup="filename=/dev/nullb0"
+		else
+			if [ "$testfile" = 'y' ]; then
```
Comment (rli9) [inline setup/fio-setup-basic:214]: thanks, here prefer to use elif

File: setup/fs [hunk 8]
```diff
@@ -115,13 +115,21 @@ probe_filesystem()
 	grep -w $fs /proc/filesystems || log_cmd modprobe $fs
 }
 
+is_null_blk()
+{
```
Comment (rli9) [inline setup/fs:125]: any common place we can put this function to avoid duplication?

Comment (lukazlim) [reply setup/fs:125]: i'm thinking of putting is_mount_on_root and is_null_blk in lib/fs_ext.sh to avoid duplication for those 2 functions, but i have to add '. $LKP_SRC/lib/fs_ext.sh' to top of fio-setup-basic, will that be fine?

File: setup/fio-setup-basic [hunk 9]
```diff
@@ -184,13 +185,18 @@ no=0
...
 	if [ "$raw_disk" = 1 ]; then
 		storage_setup="filename=$storage"
-	elif [ "$testfile" = 'y' ]; then
-		storage_setup="\
+	else
+		dev=/dev/$(basename $storage)
+		if is_null_blk $dev; then
+			storage_setup="filename=/dev/nullb0"
+		elif [ "$testfile" = 'y' ]; then
+			storage_setup="\
```
Comment (rli9) [inline setup/fio-setup-basic:203]: i propose code like below, can it work?

if [ "$raw_disk" = 1 ]; then
  storage_setup="filename=$storage"
elif is_null_blk /dev/$(basename $storage); then
  storage_setup="filename=/dev/nullb0"
elif

File: setup/fio-setup-basic [hunk 10]
```diff
@@ -189,6 +189,16 @@ for storage in $storages; do
 		dev=/dev/$(basename $storage)
 		if is_null_blk $dev; then
 			storage_setup="filename=/dev/nullb0"
+		elif is_mount_on_root $dev; then
+			[ -d "/lkp/benchmarks/fio" ] || mkdir /lkp/benchmarks/fio
```
Comment (rli9) [inline setup/fio-setup-basic:193]: refer to comment for early commit, here i think we can do below for now to keep  one series of if/elif clause.

elif is_mount_on_root /dev/$(basename $storage)

Comment (lukazlim) [reply setup/fio-setup-basic]: thanks for the feedback, i have made the requested changes

**Comment by rli9** [comment]
thanks for the patch

**Comment by lukazlim** [comment]
thanks for the feedback about all the needed changes
